src/compute/src/logging/persist.rs

```diff
@@ -0,0 +1,165 @@
+use timely::dataflow::operators::generic::builder_rc::OperatorBuilder;
+use timely::dataflow::Scope;
+use timely::PartialOrder;
+
+use mz_repr::{Diff, Row, Timestamp};
+use mz_timely_util::operators_async_ext::OperatorBuilderExt;
+use timely::progress::Antichain;
+use timely::progress::Timestamp as TimelyTimestamp;
+
+use crate::compute_state::ComputeState;
+
+// TODO(teskje): remove code duplication with `PersistSinkConnection::render_continous_sink`
+pub(crate) fn persist_sink<G>(
+    target_id: GlobalId,
+    target: &CollectionMetadata,
+    compute_state: &mut ComputeState,
+    collection: &Collection<G, Row, Diff>,
+) where
+    G: Scope<Timestamp = Timestamp>,
+{
+    let scope = collection.scope();
+    // TODO(teskje): vary active_worker_index, to distribute work for multiple sinks
+    let active_worker_index = 0;
+
+    let location = target.persist_location.clone();
+
+    let handles: Option<(WriteHandle<_, _, _, _>, _)> = if active_worker_index == scope.index() {
+        let shard_id = target.data_shard;
+        let persist_client = futures_executor::block_on(async {
+            compute_state
+                .persist_clients
+                .lock()
+                .await
+                .open(location)
+                .await
+        });
+
+        let persist_client = persist_client.expect("Successful connection");
+
+        let (write, read) = futures_executor::block_on(
+            persist_client.open::<SourceData, (), Timestamp, Diff>(shard_id),
+        )
+        .expect("could not open persist shard");
+
+        Some((write, read))
+    } else {
+        None
+    };
+
+    let mut sink = OperatorBuilder::new("Logging Persist Sink".into(), scope.clone());
+    let mut input = sink.new_input(
+        &collection.inner,
+        Exchange::new(move |_| active_worker_index as u64),
+    );
+
+    let shared_frontier = Rc::new(RefCell::new(Antichain::from_elem(
+        TimelyTimestamp::minimum(),
+    )));
+
+    compute_state
+        .reported_frontiers
+        .insert(target_id, Antichain::from_elem(0));
+
+    compute_state
+        .sink_write_frontiers
+        .insert(target_id, Rc::clone(&shared_frontier));
+
+    sink.build_async(
+        scope,
+        move |_capabilities, frontiers, scheduler| async move {
+            let (mut write, read) = match handles {
+                Some(w) => w,
+                None => return,
+            };
+
+            // Delete existing data in shard, by reading at recent_upper - 1
+            // and writing -1s at recent_upper.
+            let recent_upper = write.fetch_recent_upper().await;
+            // If this is true, we can obtain a snapshot. Otherwise, we assume the shard is empty
+            if recent_upper[0] > 0 {
+                let mut initial_flush: Vec<((SourceData, _), _, i64)> = vec![];
+                // Flush the shard
+                let since = Antichain::from_elem(recent_upper[0] - 1);
+                match read.snapshot(since).await {
```

Comment:
reading an entire snapshot at operation construction time is surprising to me. as far as I understand, operator construction is expected to be fast